core/eolearn/core/utils/parsing.py

```diff
@@ -13,15 +13,32 @@
 """
 from __future__ import annotations
 
+import sys
 from itertools import repeat
-from typing import TYPE_CHECKING, Iterable, List, Optional, Sequence, Tuple, Union, cast
+from typing import TYPE_CHECKING, Dict, Iterable, List, Optional, Sequence, Tuple, Union, cast
 
 from ..constants import FeatureType
+from .types import EllipsisType
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal
+else:
+    from typing import Literal  # pylint: disable=ungrouped-imports
 
 if TYPE_CHECKING:
     from ..eodata import EOPatch
 
 
+SingleFeatureSpec = Union[
+    Literal[FeatureType.TIMESTAMP, FeatureType.BBOX], Tuple[FeatureType, str], Tuple[FeatureType, str, str]
+]
+SequenceFeatureSpec = Sequence[Union[SingleFeatureSpec, Tuple[FeatureType, EllipsisType]]]
+
+DictFeatureSpec = Dict[FeatureType, Union[None, EllipsisType, Iterable[Union[str, Tuple[str, str]]]]]
+
+MultiFeatureSpec = Union[EllipsisType, Tuple[FeatureType, EllipsisType], SequenceFeatureSpec, DictFeatureSpec]
```

Comment:
I would keep the names, except perhaps for:
- `SingleFeatureSpec` -> `FeatureSpec` ?

I'd say it's obvious it's a single feature thing even when it's used side by side with `MultiFeatureSpec`.

Reply:
I was avoiding that name because of `eo-grow` :see_no_evil: Because in most `eo-learn` uses you need to provide input for the feature parser which is the `Union[single, multi]`. Also another case for why we'd add the `AnyFeatureSpec` (perhaps just named `FeaturesSpec`? Names are stupid).

Reply:
Good point about `eo-grow` - unless it's changed there, since it's less widely used than `eo-learn` at this point?

I like `FeaturesSpec`, I'm only skeptical about potentially misreading it, but typing is anyway meant to limit the users stupidity, not something which they would use themselves.

Reply:
I was thinking of having `FeatureSpec` as it is in eo-grow, but the full type is `FeaturesSpecification` to avoid confusion a bit